drivers/display/display_st7735r.c

```diff
@@ -0,0 +1,613 @@
+/*
+ * Copyright (c) 2020 Kim Bøndergaard <kim@fam-boendergaard.dk>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT sitronix_st7735r
+
+#include "display_st7735r.h"
+
+#include <device.h>
+#include <drivers/spi.h>
+#include <drivers/gpio.h>
+#include <sys/byteorder.h>
+#include <drivers/display.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(display_st7735r, CONFIG_DISPLAY_LOG_LEVEL);
+
+#define ST7735R_RESET_TIME              K_MSEC(1)
+#define ST7735R_EXIT_SLEEP_TIME K_MSEC(120)
+
+#define ST7735R_PIXEL_SIZE 2u
+
+struct st7735r_gpio_data {
+	const char *const name;
+	gpio_dt_flags_t flags;
+	gpio_pin_t pin;
+};
+
+struct st7735r_config {
+	const char *spi_name;
+	const char *cs_name;
+	struct spi_config spi_config;
+	struct st7735r_gpio_data cmd_data;
+	struct st7735r_gpio_data reset;
+	uint16_t height;
+	uint16_t width;
+	uint8_t madctl;
+	uint8_t colmod;
+	uint8_t caset[4];
+	uint8_t raset[4];
+	uint8_t vmctr1;
+	uint8_t invctr;
+	uint8_t pwctr1[3];
+	uint8_t pwctr2[1];
+	uint8_t pwctr3[2];
+	uint8_t pwctr4[2];
+	uint8_t pwctr5[2];
+	uint8_t frmctr1[3];
+	uint8_t frmctr2[3];
+	uint8_t frmctr3[6];
+	uint8_t gamctrp1[16];
+	uint8_t gamctrn1[16];
+};
+
+struct st7735r_data {
+	const struct st7735r_config *config;
+	const struct device *spi_dev;
+	struct spi_cs_control cs_ctrl;
+	const struct device *cmd_data_dev;
+	const struct device *reset_dev;
+	uint16_t x_offset;
+	uint16_t y_offset;
+#ifdef CONFIG_DEVICE_POWER_MANAGEMENT
+	uint32_t pm_state;
+#endif
+};
+
+static void st7735r_set_lcd_margins(struct st7735r_data *data,
+				    uint16_t x_offset, uint16_t y_offset)
+{
+	data->x_offset = x_offset;
+	data->y_offset = y_offset;
+}
+
+static void st7735r_set_cmd(struct st7735r_data *data, int is_cmd)
+{
+	gpio_pin_set(data->cmd_data_dev, data->config->cmd_data.pin, is_cmd);
+}
+
+static int st7735r_transmit(struct st7735r_data *data, uint8_t cmd,
+			    const uint8_t *tx_data, size_t tx_count)
+{
+	struct spi_buf tx_buf = { .buf = &cmd, .len = 1 };
+	struct spi_buf_set tx_bufs = { .buffers = &tx_buf, .count = 1 };
+	int ret;
+
+	st7735r_set_cmd(data, 1);
+	ret = spi_write(data->spi_dev, &data->config->spi_config, &tx_bufs);
+	if (ret < 0) {
+		return ret;
+	}
+	if (tx_data != NULL) {
+		tx_buf.buf = (void *)tx_data;
+		tx_buf.len = tx_count;
+		st7735r_set_cmd(data, 0);
+		ret = spi_write(data->spi_dev, &data->config->spi_config, &tx_bufs);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+	return 0;
+}
+
+static int st7735r_exit_sleep(struct st7735r_data *data)
+{
+	int ret;
+
+	ret = st7735r_transmit(data, ST7735R_CMD_SLEEP_OUT, NULL, 0);
```

Comment:
error code should be checked before sleeping, i.e.

```c
ret = f()
if (ret < 0) {
    return ret;
}

k_sleep(...)

return 0;
```

Reply:
In the rare situation we see an error it is probably not a problem to wait the few ms. 
But yes it was me being lazy - fixed now